### DISCUSSION TITLE
Mixins, Protocol and calling super()

### DISCUSSION THREAD

**Discussion by medihack** [microsoft/pyright#6343] [General]
From what I found, it seems that `Protocol` is the preferred way to add type hints to mixins (using them with Django). Is there a way to define `super` calls with the help of a `Protocol`? Something like this doesn't seem to work:

```python
class Base(Protocol):
    get_data: Callable

class MyMixin:
    def get(self: Base):
        self.get_data() # this is ok
        super().get_data()  # type error here
```

If I understand the issue from mypy correctly (or maybe not), then something like this [was implemented there](https://github.com/python/mypy/issues/12344).

**Comment by erictraut** [comment]
I think this is a reasonable enhancement request. It's bending the rules of protocols somewhat, but I think it's justified to handle this case.

**Comment by medihack** [comment]
Sounds good, but I wonder if something like this is more precise:

```python
class Base1(Protocol):
    get_context: Callable

class Base2(Base1, Protocol):
    get_data: Callable

class MyMixin:
    def get(self: Base2):
        self.get_data()
        super().getContext()
```

**Reply by erictraut** [reply]
Both of the above code samples will type check without error in the next release of pyright.

**Reply by medihack** [reply]
Cool! Thank you!

**Reply by medihack** [reply]
@erictraut Unfortunately, this broke again in one of the latest releases (using v1.1.348 currently) when using `...` as the method body. When calling `super` I now end up with an error `Method "xxx" cannot be called because it is abstract and unimplemented.`

```python
class Base(Protocol):
    def get_data(self):
        ...

class MyMixin:
    def get(self: Base):
        self.get_data() # this is ok
        super().get_data()  # type error here
```

**Reply by erictraut** [reply]
This has been reported [here](https://github.com/microsoft/pyright/issues/7105). I'll work on a fix for the next release.

**Reply by medihack** [reply]
@erictraut I just upgraded to v1.1.349. Unfortunately, the error remains. My last example still shows `error: Method "get_data" cannot be called because it is abstract and unimplemented (reportAbstractUsage)` when running pyright from the console.

**Reply by erictraut** [reply]
Please file a bug report so it is tracked.
